Brokerages/WebSocketClientWrapper.cs

```diff
@@ -0,0 +1,265 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.IO;
+using System.Net;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using QuantConnect.Logging;
+using QuantConnect.Util;
+
+namespace QuantConnect.Brokerages
+{
+    /// <summary>
+    /// Wrapper for System.Net.Websockets.ClientWebSocket to enhance testability
+    /// </summary>
+    public class WebSocketClientWrapper : IWebSocket
+    {
+        private const int ReceiveBufferSize = 8192;
+
+        private string _url;
+        private CancellationTokenSource _cts;
+        private ClientWebSocket _client;
+        private Task _taskConnect;
+        private readonly object _locker = new object();
+
+        /// <summary>
+        /// Static constructor for the <see cref="WebSocketClientWrapper"/> class
+        /// </summary>
+        static WebSocketClientWrapper()
+        {
+            // NET 4.5.2 and below does not enable these more secure protocols by default, so we add them in here
+            ServicePointManager.SecurityProtocol |= SecurityProtocolType.Tls11 | SecurityProtocolType.Tls12;
+        }
+
+        /// <summary>
+        /// Wraps constructor
+        /// </summary>
+        /// <param name="url"></param>
+        public void Initialize(string url)
+        {
+            _url = url;
+        }
+
+        /// <summary>
+        /// Wraps send method
+        /// </summary>
+        /// <param name="data"></param>
+        public async void Send(string data)
+        {
+            var buffer = new ArraySegment<byte>(Encoding.UTF8.GetBytes(data));
+            await _client.SendAsync(buffer, WebSocketMessageType.Text, true, _cts.Token);
+        }
+
+        /// <summary>
+        /// Wraps Connect method
+        /// </summary>
+        public void Connect()
+        {
+            lock (_locker)
+            {
+                if (!IsOpen)
```

Comment:
I think we should be checking if `_taskConnect` is already running or maybe `_cts` is null or `canceled`, so two consecutive calls to `Connect()` don't start two clients and tasks, because there can be a time gap between calling `Connect()` and `IsOpen` being true